mars/oscar/backends/mars/ray.py

```diff
@@ -0,0 +1,275 @@
+        await self.actor_pool.__on_ray_recv__(channel_id, message)
+
+    def health_check(self):
+        return PoolStatus.HEALTHY
+
+
+class RayMainPool(RayPoolBase):
+    actor_pool: RayMainActorPool
+
+    async def start(self, address, n_process, **kwargs):
+        self.actor_pool = await create_actor_pool(
+            address, n_process=n_process, pool_cls=RayMainActorPool,
+            subprocess_start_method="ray", **kwargs)
+
+
+class RaySubPool(RayPoolBase):
+    actor_pool: RaySubActorPool
+
+    async def start(self, *args, **kwargs):
+        actor_config, process_index = args
+        env = actor_config.get_pool_config(process_index)['env']
+        if env:
+            os.environ.update(env)
+        pool = await RaySubActorPool.create({
+            'actor_pool_config': actor_config,
+            'process_index': process_index
+        })
+        await pool.start()
+        self.actor_pool = pool
+        asyncio.create_task(pool.join())
+
+
+def address_to_placement_info(address):
+    """
+    Args:
+        address: The address of an actor pool which running in a ray actor. It's also
+        the name of the ray actor. address ex: ray://${pg_name}/${bundle_index}/${process_index}
+
+    Returns:
+        A tuple consisting of placement group name, bundle index, process index.
+
+    """
+    parsed_url = urlparse(unquote(address))
+    if parsed_url.scheme != "ray":
+        raise ValueError(f"The address scheme is not ray: {address}")
+    # os.path.split will not handle backslashes (\) correctly,
+    # so we use the posixpath.
+    parts = []
+    if parsed_url.netloc:
+        tmp = parsed_url.path
+        while tmp and tmp != "/":
+            tmp, item = posixpath.split(tmp)
+            parts.append(item)
+    parts = list(reversed(parts))
+    if parts and len(parts) != 2:
+        raise ValueError(f"Only bundle index and process index path are allowed in ray "
+                         f"address {address}.")
+    name, bundle_index, process_index = [parsed_url.netloc] + parts if parts else ""
+    if bool(name) != bool(bundle_index) or bool(bundle_index) != bool(process_index):
+        raise ValueError(f"Missing placement group name or bundle index or process index "
+                         f"from address {address}")
+    if name and bundle_index:
+        return name, int(bundle_index), int(process_index)
+    else:
+        return name, -1, -1
+
+
+def pg_bundle_to_address(pg_name: str, bundle_index: int, process_index: int = 0):
+    return f"ray://{pg_name}/{bundle_index}/{process_index}"
+
+
+class NodeResourceSpec:
+    __slots__ = 'n_process', 'num_cpus', 'memory', 'num_gpus'
+
+    def __init__(self, n_process, num_cpus=None, num_gpus=None, memory=None):
+        self.n_process = n_process
+        self.num_cpus = num_cpus or n_process
+        assert n_process <= self.num_cpus
+        self.num_gpus = num_gpus
+        self.memory = memory
```

Comment:
This `NodeResourceSpec` type does not support custom resource well, and it's not consistent with ray's placement group API https://docs.ray.io/en/master/placement-group.html#starting-a-placement-group

Reply:
The API will be refined when `deploy on ray` is finished.

Reply:
I think it's OK to use a plain dict to represent a bundle resource. Then we just put the dict to ray's placement group API.